domain/cloudimagemetadata/service/service.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/collections/set"
+
+	"github.com/juju/juju/domain/cloudimagemetadata"
+	"github.com/juju/juju/domain/cloudimagemetadata/errors"
+	jujuerrors "github.com/juju/juju/internal/errors"
+)
+
+// State is an interface of the persistence layer for managing cloud image metadata
+type State interface {
+	// SaveMetadata saves a list of cloud image metadata to the state.
+	SaveMetadata(ctx context.Context, metadata []cloudimagemetadata.Metadata) error
+
+	// DeleteMetadataWithImageID removes the metadata associated with the given imageID from the state.
+	DeleteMetadataWithImageID(ctx context.Context, imageID string) error
+
+	// FindMetadata retrieves a list of cloud image metadata based on the provided criteria.
+	// It returns the matched metadata or a [github.com/juju/juju/domain/cloudimagemetadata/errors.NotFound] error
+	// If there are no one.
+	FindMetadata(ctx context.Context, criteria cloudimagemetadata.MetadataFilter) ([]cloudimagemetadata.Metadata, error)
+
+	// AllCloudImageMetadata retrieves all cloud image metadata from the state and returns them as a list.
+	AllCloudImageMetadata(ctx context.Context) ([]cloudimagemetadata.Metadata, error)
+
+	// SupportedArchitectures returns a set of strings representing the architectures supported by the cloud images.
+	SupportedArchitectures(ctx context.Context) set.Strings
+}
+
+// Service provides the API for working with cloud image metadata
+type Service struct {
+	st State
+}
+
+// NewService returns a new service around the persistence layer
+func NewService(st State) *Service {
+	return &Service{
+		st: st,
+	}
+}
+
+// SaveMetadata saves the provided cloud image metadata if non-empty and valid.
+// It returns a [errors.NothingToSave] if the input are empty, or a [errors.NotValid] if at least one of the inputs are
```

Comment:
question: where is error NothingToSave defined? Looks like the method returns nil in this case, which is preferable.

Reply:
It is not. Removed it during review, forgot to remove this part of comment. Thanks for asking, I removed it.